### ISSUE TITLE
Intermittent SEGFAULT in syslog TCP endpoint

### ISSUE THREAD

**Issue by anandsinghkunwar** [fluent/fluent-bit#781]
## Bug Report

**Describe the bug**
Fluentbit encounters intermittent SEGFAULT when getting logs in syslog TCP endpoint  

**To Reproduce**
Use the syslog endpoint in TCP mode in version 0.14.2 using the docker image fluent/fluent-bit:0.14.2
```
[engine] caught signal (SIGSEGV)
Fluent-Bit v0.14.2
Copyright (C) Treasure Data

#0  0x7f24e75930c1      in  ???() at ???:0
#1  0x55e272398cc4      in  consume_bytes() at plugins/in_syslog/syslog_prot.c:30
#2  0x55e272398fc5      in  syslog_prot_process() at plugins/in_syslog/syslog_prot.c:112
#3  0x55e272398285      in  syslog_conn_event() at plugins/in_syslog/syslog_conn.c:74
#4  0x55e272360060      in  flb_engine_start() at src/flb_engine.c:542
#5  0x55e2722ffac6      in  main() at src/fluent-bit.c:824
#6  0x7f24e748a2e0      in  ???() at ???:0
#7  0x55e2722fdf49      in  ???() at ???:0
#8  0xffffffffffffffff  in  ???() at ???:0
```
- Steps to reproduce the problem:

**Expected behavior**
Expected behavior is that no seg-fault occurs.

**Your Environment**
* Version used: 0.14.2
* Configuration: 
```
  fluent-bit.conf: |
    [SERVICE]
      Flush         1
      Log_Level     debug
      Daemon        off
      Parsers_File  parsers.conf
      HTTP_Server   On
      HTTP_Listen   0.0.0.0
      HTTP_Port     2020

    # List of include files that have one of input/filter/output in them.
    @INCLUDE input-*.conf
    @INCLUDE filter-*.conf
    @INCLUDE output-*.conf
    
  input-syslog.conf: |-
    [INPUT]
      Name        syslog
      Tag         syslog
      Listen      0.0.0.0
      Port        5140
      Mode        tcp
      Parser      syslog-rfc5424-modified
      Chunk_Size  32
      Buffer_Size 1024
    
  filter-syslog.conf: |-
    [FILTER]
      Name    modify
      Match   syslog
      Rename  time @timestamp
    
    [FILTER]
      Name    lua
      Match   syslog
      script  /fluent-bit/etc/syslog.lua
      call    add_kafka_key
    
  syslog.lua: |-
    function add_kafka_key(tag, timestamp, record)
      -- Adding kafka key for syslog record
      new_record = record
      new_record["kafka_key"] = "syslog-" .. record["component"]
      return 1, timestamp, new_record
    end

    [OUTPUT]
      Name           kafka
      Match          syslog
      Brokers        server1.com:9092,server2.com:9092,server3:9092
      Topics         syslog-default,syslog-example1,syslog-example2
      Topic_Key      kafka_key
      Timestamp_Key  @timestamp
      Retry_Limit    false
      rdkafka.log.connection.close false
      rdkafka.queue.buffering.max.kbytes 10240
      rdkafka.request.required.acks 1
    
  parsers.conf: |-
    [PARSER]
      Name        syslog-rfc5424-modified
      Format      regex
      Regex       ^\<(?<priority>[0-9]{1,5})\>1 (?<time>[^ ]+) (?<host>[^ ]+) (?<component>[^ ]+) (?<subcomponent>[^ ]+) (?<pid>[-0-9]*) (?<msgid>[^ ]+) (?<extradata>(\[(.*)\]|-))[ ]+(?<message>.+)$
      Time_Key    time
      Time_Format %Y-%m-%dT%H:%M:%S.%L%z
      Time_Keep   On
      Types       pid:integer
```

* Environment name and version (e.g. Kubernetes? What version?): Kubernetes 1.10

**What I have tried**
* I have tried changing `Buffer_Size 1024` to bigger values, still doesn't help. 
* I have removed filters and outputted to stdout, still doesn't help.

**Hunch**

What I think may be happening is fluentbit crashing due to an incorrect pattern.

**Comment by newtoncorp** [comment]
I've reproduced this while trying to reproduce other issues.

parser.conf :
`[PARSER]
    Name   bug-long-input
    Format regex
    Regex  ^\<(?<pri>[0-9]+)\>(?<datetime>[^ ]* {1,2}[^ ]* [^ ]*) (?<host>[^ ]+) (?<ident>[a-zA-Z0-9_\/\.\-]*)(?:\[(?<pid>[0-9]+)\]): (?<client_ip>(?:((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?|(?<![0-9])(?:(?:[0-1]?[0-9]{1,2}|2[0-4][0-9]|25[0-5])[.](?:[0-1]?[0-9]{1,2}|2[0-4][0-9]|25[0-5])[.](?:[0-1]?[0-9]{1,2}|2[0-4][0-9]|25[0-5])[.](?:[0-1]?[0-9]{1,2}|2[0-4][0-9]|25[0-5]))(?![0-9]))):(?<client_port>(?:[+-]?(?:[0-9]+))) \[(?<accept_date>(?<haproxy_monthday>(?:(?:0[1-9])|(?:[12][0-9])|(?:3[01])|[1-9]))/(?<haproxy_month>\b(?:[Jj]an(?:uary|uar)?|[Ff]eb(?:ruary|ruar)?|[Mm](?:a|<C3><A4>)?r(?:ch|z)?|[Aa]pr(?:il)?|[Mm]a(?:y|i)?|[Jj]un(?:e|i)?|[Jj]ul(?:y)?|[Aa]ug(?:ust)?|[Ss]ep(?:tember)?|[Oo](?:c|k)?t(?:ober)?|[Nn]ov(?:ember)?|[Dd]e(?:c|z)(?:ember)?)\b)/(?<haproxy_year>(?>\d\d){1,2}):(?<haproxy_time>(?!<[0-9])(?<haproxy_hour>(?:2[0123]|[01]?[0-9])):(?<haproxy_minute>(?:[0-5][0-9]))(?::(?<haproxy_second>(?:(?:[0-5]?[0-9]|60)(?:[:.,][0-9]+)?)))(?![0-9])).(?<haproxy_milliseconds>(?:[+-]?(?:[0-9]+))))\] (?<frontend_name>\S+) (?<backend_name>\S+)/(?<server_name>\S+) (?<time_request>(?:[+-]?(?:[0-9]+)))/(?<time_queue>(?:[+-]?(?:[0-9]+)))/(?<time_backend_connect>(?:[+-]?(?:[0-9]+)))/(?<time_backend_response>(?:[+-]?(?:[0-9]+)))/(?<time_duration>\S+) (?<return_code>(?:[+-]?(?:[0-9]+))) (?<bytes_read>\S+) (?<captured_request_cookie>.*?) (?<captured_response_cookie>.*?) (?<termination_state>\S+) (?<actconn>(?:[+-]?(?:[0-9]+)))/(?<feconn>(?:[+-]?(?:[0-9]+)))/(?<beconn>(?:[+-]?(?:[0-9]+)))/(?<srvconn>(?:[+-]?(?:[0-9]+)))/(?<retries>\S+) (?<srv_queue>(?:[+-]?(?:[0-9]+)))/(?<backend_queue>(?:[+-]?(?:[0-9]+))) (\{(?<captured_request_headers>.*?)\})?( )?(\{(?<captured_response_headers>.*?)\})?( )?"(<BADREQ>|((?<http_verb>\b\w+\b) ((?<http_proto>[A-Za-z]([A-Za-z0-9+\-.]+)+)://)?(?:(?<http_user>[a-zA-Z0-9._-]+)(?::[^@]*)?@)?(?:(?<http_host>(.*)(?::(?<port>\b(?:[1-9][0-9]*)\b))?))?(?:(?<http_request>(?:/[A-Za-z0-9$.+!*'(){},~:;=@#%&_\-]*)+(?:\?[A-Za-z0-9$.+!*'|(){},~@#%&/=:;_?\-\[\]<>]*)?))?( HTTP/(?<http_version>(?:(?<![0-9.+-])(?>[+-]?(?:(?:[0-9]+(?:\.[0-9]+)?)|(?:\.[0-9]+))))))?))?"?$`

test.conf :
`[SERVICE]
    Flush        1
    Parsers_File bug-long-input.conf

[INPUT]
    Name     syslog
    Parser   bug-long-input
    Listen   127.0.0.1
    Port     5150
    Mode     tcp


[OUTPUT]
    Name     es
    Host 127.0.0.1
    Port 9200
    Index fluentd
    Type  fluentd
    Logstash_Format On
    Logstash_Prefix fluentd
    Match *

[OUTPUT]
    Name     stdout
    Match    *`

Start fluent-bit head with this configuration then create a file with a lot of entries :
`wc test-long-input
   4752   95040 1858032 test-long-input`

with log lines from haproxy or the following anonymised input :
`<142>Sep 20 16:18:44 xxxxxxx-xxx-sxxxxxx haproxy[30400]: 11.11.11.11:58150 [20/Sep/2018:16:18:44.316] xxxxxxxxx~ xxxx-xx-xxxxxxx/xxxxxxx-xxx-x01x02xx08 0/0/1/34/35 204 333 - - ---- 8/8/7/1/0 0/0 "DELETE /xxxxxxx/xxx/xxxxxxx/xxxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx?xxxxxxxxxxxxxx=xxxxxxxxxxxxxxxxxxxx&xxxxxxxxx=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx&xxxxxxx=xxxxxxxxxx HTTP/1.1"

and running the following command :
`cat test-long-input | nc -v 127.0.0.1 5150`

**Comment by edsiper** [comment]
@newtoncorp is this related to the recently fixed #774 ?

**Comment by newtoncorp** [comment]
I don't think so. All my tests where using udp except for this one on which I stumbled. Still able to reproduce.

Here's a trace with debug
```

Thread 1 "fluent-bit" received signal SIGSEGV, Segmentation fault.
__memmove_ssse3 () at ../sysdeps/x86_64/multiarch/memcpy-ssse3.S:2818
2818    ../sysdeps/x86_64/multiarch/memcpy-ssse3.S: No such file or directory.
(gdb) thread apply all bt

Thread 2 (Thread 0x7ffff6eb2700 (LWP 7175)):
#0  0x00007ffff6fbaa13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
#1  0x000000000065ebfb in _mk_event_wait (loop=0x955210) at /root/fluent-bit/lib/monkey/mk_core/mk_event_epoll.c:343
#2  0x000000000065eeeb in mk_event_wait (loop=0x955210) at /root/fluent-bit/lib/monkey/mk_core/mk_event.c:163
#3  0x0000000000428b7a in log_worker_collector (data=0x96de50) at /root/fluent-bit/src/flb_log.c:122
#4  0x0000000000438f75 in step_callback (data=0x96d7c0) at /root/fluent-bit/src/flb_worker.c:50
#5  0x00007ffff7bc16ba in start_thread (arg=0x7ffff6eb2700) at pthread_create.c:333
#6  0x00007ffff6fba41d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109

Thread 1 (Thread 0x7ffff7fe9940 (LWP 7174)):
#0  __memmove_ssse3 () at ../sysdeps/x86_64/multiarch/memcpy-ssse3.S:2818
#1  0x000000000046ae60 in consume_bytes (
    buf=0x9894e0 "f\354\221^\335Q?\333s\017\255\004X\t\335\004\002\225\n>\323vߦ\020\036\200=\350ͤd\321\063ǒ\307\342ND\343\t\311N\302]\207\016\022\236\277\017\311\005\020\336z\243\261<\362?\245\252'y\255\061}\037\375\374\031i\305ݹ?|\315ƴ\302\060\036~n\222\327\177avZ\217\353\225M\274\021n!|Y7\231\320\006\274\371\006m2\026\245\331i\250\341\334<\200\036`Q\334\327T!\036\312bwO\372؏\263+:\rxr\311hAZGJ£\353\032\327\n\253<2U\310\n\021\234\337t\326\360@\025\035ȹ\217\265\066ů\370\"\270\312\035\363ֶ\031\017\237aej\352t\310|\217\303O]e\202\037\331\r\211\332ur\373\357\361Gg\023\263\310\321\031\210'&\232\231y\177\036\344,?{\356\361\336M\213\226\227\303\325?|\033#\355\244\263\035\026rCK \341Y~\302\350\255&\277\242", <incomplete sequence \367>, bytes=53567, length=32767) at /root/fluent-bit/plugins/in_syslog/syslog_prot.c:30
#2  0x000000000046b170 in syslog_prot_process (conn=0xb227e0) at /root/fluent-bit/plugins/in_syslog/syslog_prot.c:112
#3  0x000000000046a468 in syslog_conn_event (data=0xb227e0) at /root/fluent-bit/plugins/in_syslog/syslog_conn.c:74
#4  0x0000000000434dbe in flb_engine_start (config=0x950010) at /root/fluent-bit/src/flb_engine.c:542
#5  0x00000000004208cf in main (argc=3, argv=0x7fffffffe618) at /root/fluent-bit/src/fluent-bit.c:824
(gdb)
```

We can see the the `consume_bytes` is trying to read garbage.

**Comment by edsiper** [comment]
@newtoncorp I suspect the bug is before that call in some memory corruption. Do u have a reproducible case that I can try?

**Comment by newtoncorp** [comment]
I explained how to trigger it here : https://github.com/fluent/fluent-bit/issues/781#issuecomment-423507534

Use the configuration in the comment then :

```
for i in {1..4700}; do echo '<142>Sep 20 16:18:44 xxxxxxx-xxx-sxxxxxx haproxy[30400]: 11.11.11.11:58150 [20/Sep/2018:16:18:44.316] xxxxxxxxx~ xxxx-xx-xxxxxxx/xxxxxxx-xxx-x01x02xx08 0/0/1/34/35 204 333 - - ---- 8/8/7/1/0 0/0 "DELETE /xxxxxxx/xxx/xxxxxxx/xxxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx?xxxxxxxxxxxxxx=xxxxxxxxxxxxxxxxxxxx&xxxxxxxxx=xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx&xxxxxxx=xxxxxxxxxx HTTP/1.1"' >> test-file; done
cat test-file | nc -v 127.0.0.1 5150
```

**Comment by anandsinghkunwar** [comment]
Is there any update on this @edsiper?

**Comment by edsiper** [comment]
@anandsinghkunwar  on it.

@newtoncorp would you please upload a file of the parser ? Doing copy/paste is messing up the config file (somewhere)

**Comment by newtoncorp** [comment]
@edsiper Here are files :

http://51.15.203.134/test-long-input
http://51.15.203.134/test.conf
http://51.15.203.134/bug-long-input.conf

**Comment by cite** [comment]
I might have a similar (or perhaps even identical problem) in #845 - feel free to close if duplicate.

**Comment by abh** [comment]
Streaming logs from fastly (so I don't know what the exact input is) I get what looks like the same crash within a few minutes (on 0.14.4).

```
#0  0x7ff593f1d17d      in  ???() at ???:0
#1  0x561bdacf6403      in  consume_bytes() at plugins/in_syslog/syslog_prot.c:30
#2  0x561bdacf6704      in  syslog_prot_process() at plugins/in_syslog/syslog_prot.c:112
#3  0x561bdacf59c4      in  syslog_conn_event() at plugins/in_syslog/syslog_conn.c:74
#4  0x561bdacbd2d4      in  flb_engine_start() at src/flb_engine.c:542
#5  0x561bdac5cb00      in  main() at src/fluent-bit.c:826
#6  0x7ff593eb92e0      in  ???() at ???:0
#7  0x561bdac5af69      in  ???() at ???:0
#8  0xffffffffffffffff  in  ???() at ???:0
```


[fluent-bit-configmap.yaml.gz](https://github.com/fluent/fluent-bit/files/2497911/fluent-bit-configmap.yaml.gz)

**Comment by taer** [comment]
We're doing the exact same thing(Streaming from fastly) and having the same error.  Tried 1.0.2 and have a very similar dump:
```
Fluent Bit v1.0.2
Copyright (C) Treasure Data

[2019/01/15 20:05:42] [ info] Configuration
[2019/01/15 20:05:42] [ info]  flush time     : 5 seconds
[2019/01/15 20:05:42] [ info]  input plugins  : 
[2019/01/15 20:05:42] [ info] syslog 
[2019/01/15 20:05:42] [ info]  collectors     : 
[engine] caught signal (SIGSEGV)
#0  0x7fe84b4b30ca      in  ???() at ???:0
#1  0x564dc1e20164      in  consume_bytes() at plugins/in_syslog/syslog_prot.c:31
#2  0x564dc1e20480      in  syslog_prot_process() at plugins/in_syslog/syslog_prot.c:115
#3  0x564dc1e1fb1d      in  syslog_conn_event() at plugins/in_syslog/syslog_conn.c:74
#4  0x564dc1de7b2e      in  flb_engine_start() at src/flb_engine.c:484
#5  0x564dc1d5f95f      in  main() at src/fluent-bit.c:842
#6  0x7fe84b3aa2e0      in  ???() at ???:0
#7  0x564dc1d5dda9      in  ???() at ???:0
#8  0xffffffffffffffff  in  ???() at ???:0
```

**Comment by matt-simons** [comment]
Any progress on this issue? We're seeing the same for streaming logs from Sysdig
```
#0  0x7f59ee2c3301      in  ???() at ???:0
#1  0x5560db1aa2ff      in  consume_bytes() at plugins/in_syslog/syslog_prot.c:32
#2  0x5560db1aa61b      in  syslog_prot_process() at plugins/in_syslog/syslog_prot.c:116
#3  0x5560db1a9cb8      in  syslog_conn_event() at plugins/in_syslog/syslog_conn.c:75
#4  0x5560db175868      in  flb_engine_start() at src/flb_engine.c:514
#5  0x5560db0ecd86      in  main() at src/fluent-bit.c:853
#6  0x7f59ee1ba2e0      in  ???() at ???:0
#7  0x5560db0eb269      in  ???() at ???:0
#8  0xffffffffffffffff  in  ???() at ???:0
```
Increasing the buffer size made it occur slightly less often

**Comment by edsiper** [comment]
Issue addressed on v1.3.3 

https://fluentbit.io/announcements/v1.3.3/
